File: pkg/service/notification/notification_service.go
/*
 * @Description: 通知服务实现
 * @Author: 安知鱼
 * @Date: 2025-10-12
 */
package notification

import (
	"context"
	"fmt"

	"github.com/anzhiyu-c/anheyu-app/pkg/domain/model"
	"github.com/anzhiyu-c/anheyu-app/pkg/domain/repository"
)

// UserNotificationSettings 用户通知偏好设置
type UserNotificationSettings struct {
	AllowCommentReplyNotification bool
}

// Service 通知服务接口
type Service interface {
	// 通知类型相关
	ListNotificationTypes(ctx context.Context) ([]*model.NotificationType, error)
	GetNotificationTypeByCode(ctx context.Context, code string) (*model.NotificationType, error)

	// 用户通知配置相关
	GetUserNotificationConfigs(ctx context.Context, userID uint) ([]*model.UserNotificationConfig, error)
	GetUserNotificationConfig(ctx context.Context, userID uint, notificationTypeCode string) (*model.UserNotificationConfig, error)
	UpdateUserNotificationConfig(ctx context.Context, userID uint, config *model.UserNotificationConfig) (*model.UserNotificationConfig, error)
	BatchUpdateUserNotificationConfigs(ctx context.Context, userID uint, configs []*model.UserNotificationConfig) error
	GetUserNotificationSettings(ctx context.Context, userID uint) (*UserNotificationSettings, error)

	// 通知判断相关
	ShouldNotifyUser(ctx context.Context, userID uint, notificationTypeCode string, channel string) (bool, string, error)

	// 初始化相关
	InitializeDefaultNotificationTypes(ctx context.Context) error
	EnsureUserDefaultConfigs(ctx context.Context, userID uint) error
}

type notificationService struct {
	notificationTypeRepo repository.NotificationTypeRepository
	userConfigRepo       repository.UserNotificationConfigRepository
}

// NewNotificationService 创建通知服务
func NewNotificationService(
	notificationTypeRepo repository.NotificationTypeRepository,
	userConfigRepo repository.UserNotificationConfigRepository,
) Service {
	return &notificationService{
		notificationTypeRepo: notificationTypeRepo,
		userConfigRepo:       userConfigRepo,
	}
}

// ListNotificationTypes 获取所有通知类型
func (s *notificationService) ListNotificationTypes(ctx context.Context) ([]*model.NotificationType, error) {
	return s.notificationTypeRepo.FindAll(ctx)
}

// GetNotificationTypeByCode 根据code获取通知类型
func (s *notificationService) GetNotificationTypeByCode(ctx context.Context, code string) (*model.NotificationType, error) {
	return s.notificationTypeRepo.FindByCode(ctx, code)
}

// GetUserNotificationConfigs 获取用户的所有通知配置
func (s *notificationService) GetUserNotificationConfigs(ctx context.Context, userID uint) ([]*model.UserNotificationConfig, error) {
	// 先确保用户有默认配置
	if err := s.EnsureUserDefaultConfigs(ctx, userID); err != nil {
		return nil, err
	}

	return s.userConfigRepo.FindByUserID(ctx, userID)
}

// GetUserNotificationConfig 获取用户指定类型的通知配置
func (s *notificationService) GetUserNotificationConfig(ctx context.Context, userID uint, notificationTypeCode string) (*model.UserNotificationConfig, error) {
	return s.userConfigRepo.FindByUserAndType(ctx, userID, notificationTypeCode)
}

// UpdateUserNotificationConfig 更新用户通知配置
func (s *notificationService) UpdateUserNotificationConfig(ctx context.Context, userID uint, config *model.UserNotificationConfig) (*model.UserNotificationConfig, error) {
	config.UserID = userID
	return s.userConfigRepo.CreateOrUpdate(ctx, config)
}

// BatchUpdateUserNotificationConfigs 批量更新用户通知配置
func (s *notificationService) BatchUpdateUserNotificationConfigs(ctx context.Context, userID uint, configs []*model.UserNotificationConfig) error {
	return s.userConfigRepo.BatchCreateOrUpdate(ctx, userID, configs)
}

// GetUserNotificationSettings 获取用户通知偏好设置（简化版）
func (s *notificationService) GetUserNotificationSettings(ctx context.Context, userID uint) (*UserNotificationSettings, error) {
	// 获取用户的所有通知配置
	configs, err := s.GetUserNotificationConfigs(ctx, userID)
	if err != nil {
		return nil, fmt.Errorf("获取通知配置失败: %w", err)
	}

	// 转换为简化格式
	settings := &UserNotificationSettings{
		AllowCommentReplyNotification: false, // 默认关闭
	}

	for _, config := range configs {
		if config.NotificationType == nil {
			continue
		}

		if config.NotificationType.Code == model.NotificationTypeCommentReply {
			settings.AllowCommentReplyNotification = config.IsEnabled
			break
		}
	}

	return settings, nil
}

// ShouldNotifyUser 检查是否应该通知用户
func (s *notificationService) ShouldNotifyUser(ctx context.Context, userID uint, notificationTypeCode string, channel string) (bool, string, error) {
	return s.userConfigRepo.ShouldNotify(ctx, userID, notificationTypeCode, channel)
}

// InitializeDefaultNotificationTypes 初始化默认通知类型
func (s *notificationService) InitializeDefaultNotificationTypes(ctx context.Context) error {
	defaultTypes := model.DefaultNotificationTypes()

	for _, nt := range defaultTypes {
		// 检查是否已存在
		existing, err := s.notificationTypeRepo.FindByCode(ctx, nt.Code)
		if err != nil {
			return fmt.Errorf("检查通知类型失败: %w", err)
		}

		if existing == nil {
			// 不存在，创建
			_, err := s.notificationTypeRepo.Create(ctx, nt)
			if err != nil {
				return fmt.Errorf("创建通知类型失败 [%s]: %w", nt.Code, err)
			}
		} else {
			// 存在，更新（除了code和id外的字段）
			nt.ID = existing.ID
			_, err := s.notificationTypeRepo.Update(ctx, existing.ID, nt)
			if err != nil {
				return fmt.Errorf("更新通知类型失败 [%s]: %w", nt.Code, err)
			}
		}
	}

	return nil
}

// EnsureUserDefaultConfigs 确保用户拥有所有默认通知类型的配置
func (s *notificationService) EnsureUserDefaultConfigs(ctx context.Context, userID uint) error {
	// 获取所有通知类型
	notificationTypes, err := s.notificationTypeRepo.FindAll(ctx)
	if err != nil {
		return fmt.Errorf("获取通知类型列表失败: %w", err)
	}

	// 获取用户已有的配置
	userConfigs, err := s.userConfigRepo.FindByUserID(ctx, userID)
	if err != nil {
		return fmt.Errorf("获取用户通知配置失败: %w", err)
	}

	// 构建已有配置的map
	existingConfigsMap := make(map[uint]bool)
	for _, config := range userConfigs {
		existingConfigsMap[config.NotificationTypeID] = true
	}

	// 为缺失的通知类型创建默认配置
	for _, nt := range notificationTypes {
		if !existingConfigsMap[nt.ID] {
			config := &model.UserNotificationConfig{
				UserID:             userID,
				NotificationTypeID: nt.ID,
				IsEnabled:          nt.DefaultEnabled,
				EnabledChannels:    nt.SupportedChannels,
				CustomSettings:     make(map[string]interface{}),
			}

			_, err := s.userConfigRepo.CreateOrUpdate(ctx, config)
			if err != nil {
				return fmt.Errorf("创建用户默认通知配置失败 [type_id=%d]: %w", nt.ID, err)
			}
		}
	}

	return nil
}

File: server/src/notification/notification.service.ts
import { Injectable, Logger, NotFoundException, OnModuleInit } from '@nestjs/common';
import { NotificationRepository } from './notification.repository';
import { UpdateSimpleNotificationSettingsDto } from './dto/simple-notification-settings.dto';
import { UpdateUserNotificationConfigDto } from './dto/user-notification-config.dto';
import { ErrorCodes } from '../common/constants/error-codes';

/**
 * Default notification types per D-220.
 * Matches Go pkg/domain/model/notification.go DefaultNotificationTypes().
 */
const DEFAULT_NOTIFICATION_TYPES = [
  {
    code: 'comment_reply',
    name: '评论回复通知',
    description: '当您的评论被他人回复时通知您',
    category: 'comment',
    isActive: true,
    defaultEnabled: true,
    supportedChannels: ['email', 'push'],
  },
  {
    code: 'comment_new',
    name: '新评论通知',
    description: '当网站收到新评论时通知博主',
    category: 'comment',
    isActive: true,
    defaultEnabled: true,
    supportedChannels: ['email', 'push'],
  },
  {
    code: 'system_update',
    name: '系统更新通知',
    description: '接收系统更新和新功能介绍',
    category: 'system',
    isActive: true,
    defaultEnabled: true,
    supportedChannels: ['email'],
  },
  {
    code: 'marketing_promo',
    name: '营销推广通知',
    description: '接收活动推荐和优惠信息',
    category: 'marketing',
    isActive: true,
    defaultEnabled: false,
    supportedChannels: ['email'],
  },
];

@Injectable()
export class NotificationService implements OnModuleInit {
  private readonly logger = new Logger(NotificationService.name);

  constructor(private readonly repo: NotificationRepository) {}

  /**
   * Initialize default notification types on startup.
   * Per D-220, matches Go InitializeDefaultNotificationTypes.
   */
  async onModuleInit() {
    await this.initializeDefaultNotificationTypes();
  }

  /**
   * Per D-220, check-then-create-or-update pattern for idempotent initialization.
   * Matches Go InitializeDefaultNotificationTypes exactly.
   * Per Pitfall 7 in RESEARCH.md, this avoids UNIQUE constraint violations on restart.
   */
  async initializeDefaultNotificationTypes() {
    for (const defaultType of DEFAULT_NOTIFICATION_TYPES) {
      const existing = await this.repo.findNotificationTypeByCode(defaultType.code);
      if (!existing) {
        await this.repo.createNotificationType(defaultType);
        this.logger.log(`Created notification type: ${defaultType.code}`);
      } else {
        await this.repo.updateNotificationType(existing.id, {
          name: defaultType.name,
          description: defaultType.description,
          category: defaultType.category,
          isActive: defaultType.isActive,
          defaultEnabled: defaultType.defaultEnabled,
          supportedChannels: defaultType.supportedChannels,
        });
      }
    }
    this.logger.log('Default notification types initialized');
  }

  /**
   * Per D-220, ensure user has default configs for all notification types.
   * Matches Go EnsureUserDefaultConfigs.
   */
  async ensureUserDefaultConfigs(userId: number) {
    const allTypes = await this.repo.findNotificationTypes();
    const existingConfigs = await this.repo.findUserNotificationConfigs(userId);
    const existingTypeIds = new Set(
      existingConfigs.map((c: any) => c.notificationTypeId),
    );

    for (const type of allTypes) {
      if (!existingTypeIds.has(type.id)) {
        await this.repo.createUserNotificationConfig({
          userId,
          notificationTypeId: type.id,
          isEnabled: type.defaultEnabled,
          enabledChannels: type.supportedChannels,
        });
      }
    }

    // Return all configs after ensuring defaults
    return this.repo.findUserNotificationConfigs(userId);
  }

  /**
   * List all notification types.
   * Matches Go ListNotificationTypes.
   */
  async listNotificationTypes() {
    return this.repo.findNotificationTypes();
  }

  /**
   * Find a notification type by code.
   * Exposed for Plan 07 comment→notification integration.
   * Matches Go GetNotificationTypeByCode.
   */
  async findNotificationTypeByCode(code: string) {
    return this.repo.findNotificationTypeByCode(code);
  }

  /**
   * Get simplified notification settings for the current user.
   * Per D-220, only exposes allowCommentReplyNotification boolean.
   * Matches Go GetUserNotificationSettings.
   */
  async getUserNotificationSettings(userId: number) {
    await this.ensureUserDefaultConfigs(userId);
    const configs = await this.repo.findUserNotificationConfigs(userId);

    // Find the comment_reply type config
    const commentReplyConfig = (configs as any[]).find(
      (c: any) => c.notificationType?.code === 'comment_reply',
    );

    return {
      allowCommentReplyNotification: commentReplyConfig?.isEnabled ?? true,
    };
  }

  /**
   * Update simplified notification settings.
   * Per D-220, updates the comment_reply type config.
   * Matches Go UpdateUserNotificationSettings.
   */
  async updateUserNotificationSettings(
    userId: number,
    dto: UpdateSimpleNotificationSettingsDto,
  ) {
    // Find comment_reply type
    const commentReplyType = await this.repo.findNotificationTypeByCode('comment_reply');
    if (!commentReplyType) {
      throw new NotFoundException(ErrorCodes.NOTIFICATION_TYPE_NOT_FOUND);
    }

    // Find or create user config for comment_reply
    let config = await this.repo.findUserNotificationConfig(userId, commentReplyType.id);
    if (!config) {
      config = await this.repo.createUserNotificationConfig({
        userId,
        notificationTypeId: commentReplyType.id,
        isEnabled: dto.allowCommentReplyNotification,
        enabledChannels: commentReplyType.supportedChannels,
      });
    } else {
      config = await this.repo.updateUserNotificationConfig(config.id, {
        isEnabled: dto.allowCommentReplyNotification,
      });
    }

    return {
      allowCommentReplyNotification: dto.allowCommentReplyNotification,
    };
  }

  /**
   * Get full notification configs for the current user.
   * Per D-220, returns all configs with nested notificationType info.
   * Matches Go GetUserNotificationConfigs.
   */
  async getUserNotificationConfigs(userId: number) {
    return this.ensureUserDefaultConfigs(userId);
  }

  /**
   * Update a specific notification config.
   * Matches Go UpdateUserNotificationConfig.
   */
  async updateUserNotificationConfig(
    userId: number,
    configId: number,
    dto: UpdateUserNotificationConfigDto,
  ) {
    const updated = await this.repo.updateUserNotificationConfig(configId, dto);
    if (!updated) {
      throw new NotFoundException(ErrorCodes.NOTIFICATION_CONFIG_UPDATE_FAILED);
    }
    return updated;
  }

  /**
   * Check if a user should be notified for a given type and channel.
   * Matches Go ShouldNotifyUser.
   */
  async shouldNotifyUser(userId: number, typeCode: string, channel: string): Promise<boolean> {
    const type = await this.repo.findNotificationTypeByCode(typeCode);
    if (!type || !type.isActive) return false;

    const config = await this.repo.findUserNotificationConfig(userId, type.id);
    if (!config || !config.isEnabled) return false;

    const channels: string[] = config.enabledChannels ?? [];
    return channels.includes(channel);
  }

  /**
   * Create an in-app notification record.
   * Per D-218, creates a notification in the notifications table.
   */
  async createNotification(params: {
    userId: number;
    notificationTypeId: number;
    title: string;
    content?: string;
  }) {
    return this.repo.createNotification(params);
  }

  /**
   * List in-app notifications with pagination and optional isRead filter.
   * Per D-218.
   */
  async listNotifications(
    userId: number,
    options: { page: number; pageSize: number; isRead?: boolean },
  ) {
    const { list, total } = await this.repo.findNotifications(userId, options);
    return {
      list,
      total,
      page: options.page,
      pageSize: options.pageSize,
    };
  }

  /**
   * Mark a single notification as read.
   * Per D-218, user-scoped for security (T-09-11).
   */
  async markNotificationAsRead(notificationId: number, userId: number) {
    const notification = await this.repo.findNotificationById(notificationId, userId);
    if (!notification) {
      throw new NotFoundException(ErrorCodes.NOTIFICATION_NOT_FOUND);
    }
    return this.repo.markNotificationAsRead(notificationId);
  }

  /**
   * Mark all unread notifications as read for a user.
   * Per D-218.
   */
  async markAllNotificationsAsRead(userId: number) {
    await this.repo.markAllNotificationsAsRead(userId);
  }

  /**
   * Get unread notification count for a user.
   * Per D-218.
   */
  async getUnreadCount(userId: number) {
    const count = await this.repo.countUnreadNotifications(userId);
    return { count };
  }
}
